### DISCUSSION TITLE
How to use .clang-tidy config file?

### DISCUSSION THREAD

**Discussion by SujjithS** [clangd/clangd#1801] [Q&A]
```
root
|
--SW---------Components folder
|                    |
|                   A-----------------------X
|                    |                      |
|                    |                  CMakeList.txt(2)                           
|                   B
|                    |
|                    |
|                    |
|                   CMakeList.txt(1)
|
|
--tools--sca--clang---.clang-tidy (config file with many selective checks)
```

CMakeList.txt(1) has the below section,
```
if(CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
	option(BL_ENABLE_CLANG_TIDY_CHECKS "Run clang-tidy with the compiler." OFF)
endif()
```

CMakeList.txt(2) has the below section
```
if (BL_ENABLE_CLANG_TIDY_CHECKS)    
    include(ClangTidy)
    AddClangTidy(${LIBRARY_NAME} )
endif()
```

I am using a **cLang-tidy.cmake** in a different location whose contents are

```
find_program(CLANG-TIDY_PATH clang-tidy REQUIRED)

function(AddClangTidy target)

  set_target_properties(${target}
    PROPERTIES CXX_CLANG_TIDY
    "${CLANG-TIDY_PATH};-checks=*,\
-llvm-include-order,\
-modernize-use-trailing-return-type,\
-readability-convert-member-functions-to-static,\
-cppcoreguidelines-avoid-magic-numbers")

endfunction()
```
It works fine with the setup. But I am quite sure that I shouldnt need to use the 'checks- *' in the cLang-tidy.cmake since I already have a config file. I observe that the config file is not getting used at the moment. How should I locate the config file from this cmake?

**Comment by SujjithS** [comment]
clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 10.0.0
  
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: skylake

I have tried the below already

```
function(AddClangTidy target)

  set_target_properties(${target}
    PROPERTIES CXX_CLANG_TIDY
    "${CLANG-TIDY_PATH};--config-file=${CMAKE_SOURCE_DIR}/.clang-tidy")
```
    
    
   clang-tidy: Unknown command line argument '--config-file=/workspaces/ad-radar-sensor/software/business_logic/.clang-tidy'.  Try: '/usr/bin/clang-tidy --help'
clang-tidy: Did you mean '--config=/workspaces/Main/Components/business_logic/.clang-tidy'?



```
function(AddClangTidy target)

  set_target_properties(${target}
    PROPERTIES CXX_CLANG_TIDY
    "${CLANG-TIDY_PATH};--config=${CMAKE_SOURCE_DIR}/.clang-tidy")
```
    
    
YAML:1:1: error: not a mapping
/workspaces/Main/Components/business_logic/.clang-tidy
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error: invalid configuration specified.

**Reply by HighCommander4** [reply]
The reason for the `Unknown command line argument '--config-file=...'` error is that you're using clang-tidy 10, and the `--config-file` option was introduced in clang-tidy 12.

That said, I don't think this option is relevant to clangd in any way. Clangd searches for a `.clang-tidy` file in the ancestor directories of the source file being opened; there is no support currently for pointing clangd to a `.clang-tidy` file elsewhere (there is an issue on file about this, https://github.com/clangd/clangd/issues/805).

A possible workaround is to have the build process copy the `.clang-tidy` file to an ancestor directory of the source files it should apply to.
